### PR TITLE
Add Bounty-related events

### PR DESCRIPTION
**Pull request by zakhap**
To move forward with [this commonwealth ticket](https://github.com/hicommonwealth/commonwealth/issues/891), we need to watch Bounty events on chain. This PR adds Bounty-related events for substrate-based chains.

### CODE REVIEW

File: src/interfaces.ts [hunk 1]
```diff
@@ -204,6 +210,30 @@ export function eventToEntity(event: IChainEventKind): [ IChainEntityKind, Entit
       return [ SubstrateTypes.EntityKind.TreasuryProposal, EntityEventKind.Complete ];
     }
 
+    // Bounty Events
+    case SubstrateTypes.EventKind.TreasuryBountyProposed: {
+      return [ SubstrateTypes.EntityKind.Bounty, EntityEventKind.Create ];
+    }
+    case SubstrateTypes.EventKind.TreasuryBountyAwarded: {
+      return [ SubstrateTypes.EntityKind.Bounty, EntityEventKind.Update ]; // TODO: I believe this is correct, check frame/treasury/src/lib.rs 995-1020
```
[the remark this answers is not in the document]
Comment (jnaviask) [reply src/interfaces.ts:218]: If an entity event is `Complete`, that means no further events will be received for the same entity. It effectively "completes" the entity. Is that the case here?

[the remark this answers is not in the document]
Comment (jnaviask) [reply src/interfaces.ts:218]: I'll do a full code review later today.

[the remark this answers is not in the document]
Comment (zakhap) [reply src/interfaces.ts:218]: >  It effectively "completes" the entity. Is that the case here?

I don't think so... BountyAwarded occurs when the task is completed "done" and then BountyClaimed needs to be called by the beneficiary for the Bounty to be completed in full.

[the remark this answers is not in the document]
Comment (jnaviask) [reply src/interfaces.ts:218]: In that case, BountyAwarded is an "Update", and BountyClaimed is "Complete".

File: src/substrate/filters/enricher.ts [hunk 2]
```diff
@@ -455,6 +455,110 @@ export async function Enrich(
         };
       }
 
+      case EventKind.TreasuryBountyProposed: {
+        const [ bountyIndex ] = event.data as unknown as [ BountyIndex ] & Codec;
+        const bountyOpt = await api.query.treasury.bounties<Option<Bounty>>(bountyIndex);
+        if (!bountyOpt.isSome) {
+          throw new Error(`could not fetch treasury proposal index ${+bountyIndex}`);
```
Comment (jnaviask) [inline src/substrate/filters/enricher.ts:462]: May want to change this copy to "treasury bounty index".

File: src/substrate/filters/enricher.ts [hunk 3]
```diff
@@ -455,6 +455,110 @@ export async function Enrich(
...
+        return {
+          data: {
+            kind,
+            bountyIndex: +bountyIndex,
+            proposer: bounty.proposer.toString(),
+            value: bounty.value.toString(),
+            fee: bounty.fee.toString(),
+            curatorDeposit: bounty.curatorDeposit.toString(),
+            bond: bounty.bond.toString(),
+            status: {
```
Comment (jnaviask) [inline src/substrate/filters/enricher.ts:474]: Oof -- why are we storing the entire status here, when I assume at time of "BountyProposed" we know what state it's in?

File: src/substrate/storageFetcher.ts [hunk 4]
```diff
@@ -246,6 +250,35 @@ export class StorageFetcher extends IStorageFetcher<ApiPromise> {
...
+      const { proposer, value, fee, curatorDeposit, bond, } = bounties[index].unwrap();
+      return {
+        kind: EventKind.TreasuryBountyProposed,
+        bountyIndex: +id,
+        proposer: proposer.toString(),
+        value: value.toString(),
+        fee: fee.toString(),
+        curatorDeposit: curatorDeposit.toString(),
+        bond: bond.toString(),
+    } as ITreasuryBountyProposed;
```
Comment (jnaviask) [inline src/substrate/storageFetcher.ts:276]: We should generate additional update events here if the bounty is in a state beyond "proposed".

File: src/substrate/types.ts [hunk 5]
```diff
@@ -72,6 +98,7 @@ export enum EntityKind {
   TreasuryProposal = 'treasury-proposal',
   CollectiveProposal = 'collective-proposal',
   SignalingProposal = 'signaling-proposal',
+  Bounty = 'bounty',
```
Comment (jnaviask) [inline src/substrate/types.ts:101]: I would call it a TreasuryBounty rather than just "bounty".

**Review by jnaviask — CHANGES_REQUESTED** [review]
Generally looks good! Main change I want is to the status object in the Proposed event -- we should be able to simplify that, since we're tracking all status updates through events anyway. And then we can update that in the fetcher as well.

File: src/substrate/types.ts (hunk 5 above)
Comment (zakhap) [reply src/substrate/types.ts:101]: 👍

File: src/substrate/filters/enricher.ts (hunk 2 above)
Comment (zakhap) [reply src/substrate/filters/enricher.ts:462]: I named it after the treasury interface type:
`/** @name BountyIndex */ export interface BountyIndex extends u32 {}`

File: src/substrate/filters/enricher.ts (hunk 3 above)
Comment (zakhap) [reply src/substrate/filters/enricher.ts:474]: You're right! We can also always verify the status by querying for the bounty on-chain.

**Comment by zakhap** [comment]
@jnaviask I think this PR is ready to be reviewed "one last time" before merge. There's an issue with the fakeAPI I'll ask about in the morning with simulating the 'DeriveBounty' type, but shouldn't be more than test hygiene.

File: test/unit/edgeware/testUtil.ts [hunk 6]
```diff
@@ -183,8 +186,11 @@ export function constructFakeApi(
       },
       treasury: {
         proposals,
+        bounties,
```
Comment (jnaviask) [inline test/unit/edgeware/testUtil.ts:189]: These tests will need to be moved to derives as well once the enricher logic is changed.

**Review by jnaviask — CHANGES_REQUESTED** [review]
Few small fixes. Also, need to update type_parser.ts as well to ensure that bounty events are being properly detected based on the chain's emitted events/extrinsics.

**Review by jnaviask — COMMENTED** [review]
Code looks good! Will wait to merge until we make these small test fixes.

**Comment by jnaviask** [comment]
Looks good!

**Pull request merged**